File: js/lib/fetch.js
var widgets = require('@jupyter-widgets/base');
var _ = require('lodash');

var FetchModel = widgets.DOMWidgetModel.extend({
  defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
      _model_name: 'FetchModel',
      _view_name: 'FetchView',
      _model_module: 'ipyfetch',
      _view_module: 'ipyfetch',
      _model_module_version: '0.1.0',
      _view_module_version: '0.1.0',
      result: {},
      value: 'js'
    }
  ),
});


var FetchView = widgets.DOMWidgetView.extend({
  initialize: function () {
    this.model.on('change:url', this.render, this);
  },

  render: function () {
    let options;
    if (this.model.get('shared_origin')) {
      options = {
        mode: 'no-cors',
        credentials: 'same-origin',
        headers: new Headers({'Access-Control-Allow-Origin': '*'})
      };
    }
    const request = async () => {
      const response = await fetch(
        this.model.get('url'),
        options
      );
      const result = await response.json();
      console.log(result);
      this.model.set('value', result);
      this.touch();
    };
    request();
  },


});


module.exports = {
  FetchModel: FetchModel,
  FetchView: FetchView
};

File: ipyfetch/fetch.py
import ipywidgets as widgets
from traitlets import Unicode, Dict as DictTrait, Bool
from typing import Callable, Container, Dict


@widgets.register
class Fetch(widgets.DOMWidget):
    """Fetch widget"""

    _view_name = Unicode("FetchView").tag(sync=True)
    _model_name = Unicode("FetchModel").tag(sync=True)
    _view_module = Unicode("ipyfetch").tag(sync=True)
    _model_module = Unicode("ipyfetch").tag(sync=True)
    _view_module_version = Unicode("^0.1.0").tag(sync=True)
    _model_module_version = Unicode("^0.1.0").tag(sync=True)

    url = Unicode("").tag(sync=True)
    value = DictTrait({}).tag(sync=True)
    shared_origin = Bool(False).tag(sync=True)


def create_on_fetch(url: str, container: Dict) -> Callable:
    """
    Return a function to fetch URL.
    On fetch URL, fetch URL directly.
    It returns the result in container with key 'result'
    e.g container['result'] = <JSON_URL_RESPONSE>
    """
    f = Fetch(url=url)

    def on_fetch(change, *args, **kwargs):
        container["result"] = change["new"]

    f.observe(on_fetch, "value")
    return lambda: f
